### ISSUE TITLE
False URLs added to site tree on second spider

### ISSUE THREAD

**Issue by jhowlett96** [zaproxy/zaproxy#7161]
### Describe the bug

When spidering site, if ticked 'POST forms (recommended but may generate unwanted requests)'. On the second spider, URLs that were found with a POST request in the first spider are added again to Site tree as GET requests.

### Steps to reproduce the behavior

1. spider site
2. spider same site again so you know it will revisit a POST request.
3. see site tree and find GET request with same URL as original POST.

### Expected behavior

Do not add these erroneous GET requests

### Software versions

ZAP 2.11.1

### Would you like to help fix this issue?

- [ ] Yes

**Comment by kingthorin** [comment]
What scripts do you have enabled?
Are you sure you didn’t active scan? (The GET for POST rule might ‘seem’ to cause this.)

**Comment by jhowlett96** [comment]
I have no custom scripts enabled (I do have an input vector script on my system but this bug still occurs when it is disabled)

I can confirm that I am not doing an active scan. All I am doing is spidering the same URL twice.

 Funnily enough the GET for POST rule is turned off for the scan policy that is loaded on my session anyway.

**Comment by thc202** [comment]
That's the expected behaviour, the URLs present in the Sites tree (regardless of the method) are added as seeds when starting the spider thus later requested as GET. (Just clarifying not saying that it can't or shouldn't be changed.)

**Comment by thc202** [comment]
Tweaked the issue to match the behaviour requested.

**Comment by jhowlett96** [comment]
This also occurs to pages that the spider didn't start on. I was just stating that it also happens when I start using the same URL as before. It happens to all POST requests found.

**Comment by thc202** [comment]
Yes, as mentioned in the above comment "the URLs present in the Sites tree are added as seeds", it does not matter the method or the source of those requests (whether they were accessed manually or by the spider or other tool).

**Comment by jhowlett96** [comment]
They are removed from the exporting of the site tree and so are recognisable as 'seeds'.

Why are they present in the sites tree if they are not exported, what benefit does the user have in having these 'false' (I say false because there are no references to these URLs with these arguments as they are created by ZAP when completing multiple spiders) URLs present in the sites tree?

**Comment by jhowlett96** [comment]
Is there a way of distinguishing these seed URLs in an active scan rule? 

I am trying to develop a new active scan rule but they are causing many false positives so I would like be able to detect and then ignore them for my scan rule.

**Comment by kingthorin** [comment]
Check if the method of getBaseMsg() is GET? There might be other ways too but that seems like a quick solution while you’re testing.

**Comment by jhowlett96** [comment]
Background: looking at the feasibility of a new active CSRF scanner rule that checks response when CSRF-token value changed, to test that the second page is actually testing the CSRF-token value.

, that was my first thought but I need to test GET requests too because any type of request can modify the database, Therefore this new rule must target even GET requests. These seed URLs are lacking CSRF-tokens (as they were in the form data that was stripped while turning method from POST to GET) raising many false positives.

**Comment by kingthorin** [comment]
Check if they have any params and skip if they do not?

**Comment by jhowlett96** [comment]
I might have to go with that for now, I hoped there was a way to distinguish these URLs but alas. That might be the best solution as it will target most URLs.
The only case that condition you proposed would exclude a valid URL would be a database edit that just cycles between values so a new value wouldn't need to be given in the request for example putting a site into maintenance mode (which would be a very bad page to miss a CSRF token on thinking about it).

**Comment by thc202** [comment]
fwiw, they are not "seed URLs" they are URLs used as seeds, there's nothing special about them nor anything that can distinguish them.

**Comment by jhowlett96** [comment]
@thc202 What does a seed mean? I initially thought it was where ZAP would start its spider but there is only one of them for each spider being run which is set in the spider config before running

**Comment by thc202** [comment]
It's that, but as mentioned previously what's in the Sites tree is used to seed the Spider as well: https://github.com/zaproxy/zaproxy/blob/b7f2e4c8c65db1ef96bc7a3c4c5f1c04f05ae8a0/zap/src/main/java/org/zaproxy/zap/extension/spider/SpiderThread.java#L282-L310

**Comment by jhowlett96** [comment]
Why does ZAP try to use POST pages as seeds given that a POST request turned into a GET request will most likely throw up an error(because all the form data is lost) and not give the spider anything to spider to?

**Comment by thc202** [comment]
A URL can be used for any method, there's no such restriction that a POST to a URL can't handle a GET.

**Comment by jhowlett96** [comment]
yes that's true but what I'm getting at is that if the spider only found a POST with some form data as the request made to a given URL. What would be gained from making a GET request that is most likely invalid?

I can see website structures that would be using the same URLs for many different purposes and therefore a GET request from a POST request may give a valid response. I can see two sides to it and therefore I am wondering if a configurable option could be added to the spider that could stop this behaviour if wanted.

**Comment by jhowlett96** [comment]
@kingthorin Unfortunately checking if there are 0 params is not enough of a condition to distinguish these URLs. This is because the POSTs turned into GETs through the spider only strip the form data but keep the URL parameters and so they could have any number of params. 

Why is the spider turning POSTs into GETs when determining seeds for the spider? Wouldn't it be more beneficial to use the POSTs as POSTs instead to have a higher probability of returning a valid response?

**Comment by psiinon** [comment]
The POSTs should appear in the sites tree. If they dont then theres something going wrong with your setup.

**Comment by jhowlett96** [comment]
The POSTs do appear in my site tree. The problem is that GET requests are being made that are generated from the POST requests of a previous spider. These GET requests are then being targeted by the active scan. 

Why are the POST requests being turned into GET requests when generating seeds for the next spider and not left as the original POST requests?

**Comment by thc202** [comment]
To improve coverage, just because a POST was sent to a URL does not mean it's not serving GETs.

The scan rule will have the same problem regardless of the spider behaviour. The user can be proxying/sending all kind of funny things before starting the active scan.

**Comment by psiinon** [comment]
The ZAP spiders aim to explore sites as thoroughly as possible in order to scan them better. Thats how it works and its not something we are going to change.
If you want to reduce the coverage on your app then there are various options like excluding URLs, removing entries from the sites tree etc.
Also see https://blog.mozilla.org/security/2013/07/10/how-to-speed-up-owasp-zap-scans/

**Comment by jhowlett96** [comment]
Ok I thought it might of been a bug that was causing all seeds to be GETs but if that is the intended behaviour as it can increase coverage then I can't argue with that. 

Then I ask for a configurable option that could be 'use GET requests for POST requests found in site tree (recommended)' as an option that is set on by default in the spidering options.

Some people may not want these requests generated on spidering because they know their URL will only serve a POST. This behaviour generates an extra request per POST URL which could be increasing the active scan time considerably. My current scenario is that I have a scan containing 2000 URLs with roughly 15% being POSTs which would mean this behaviour is generating ~300 extra requests (that I know are invalid) that are then attacked.

**Comment by kingthorin** [comment]
Is there a reason you’re spidering twice? Maybe that’s the issue we should be addressing 🤷‍♂️

**Comment by jhowlett96** [comment]
I must spider twice because some pages have the same form action but send the user to a different page. 

A simple example, a table of users on system next to each user there is a button 'Edit' & 'Delete' they both are on the the same form so same action but one redirects to an edit specific page and the other redirects to a delete specific page. 

ZAP will go to one of these pages redirect pages but I must re-spider on the page not found to spider all pages in the user edit/delete area.

**Comment by jhowlett96** [comment]
Ok so I have done some coverage testing on my product and found that not using existing URLs as seeds has yielded a higher coverage (1870 URLS with, 1940 URLs without). Therefore I would like to ask to have a configuration option in the Spider tab of the Options Panel. 

This is the code block that I "if(false) {" for testing: 

https://github.com/zaproxy/zaproxy/blob/1a4af3a5678488a7d4f5559cb82cb3f89ae483c2/zap/src/main/java/org/zaproxy/zap/extension/spider/SpiderThread.java#L322-L326

I am happy for the default to be true. So text would be along the lines "Use existing URLs in the site tree as seeds".

If you would like me to complete this work then I am happy to do so but as it is editing the core then I can understand if you would like to take ownership of the task.

**Comment by jhowlett96** [comment]
I mentioned this issue on https://groups.google.com/g/zaproxy-develop/c/V5FD9XKkRng

**Comment by jhowlett96** [comment]
Obviously it is product based and someone else's product may behave differently.

I suspect it is because it is navigating the website in a slightly different way (URL request order has now changed as there are more URLs) and so the spider can reach certain URLs that are cut off when accessing the seeds (maybe its a delete user page for example and now the user details page cant be accessed anymore) on the subsequent spiders.
